File: src/engine-host/gpu/ipc.ts
import { readSync } from 'node:fs'
import { GPU_IPC_MAGIC, GPU_PROTOCOL_VERSION } from '@shared/gpu/protocol'

export interface GpuIpcMessage {
  header: unknown
  blobs: Map<string, Buffer>
}

export function encodeGpuMessage(header: unknown, blobs: Map<string, Buffer> = new Map()): Buffer {
  const json = Buffer.from(JSON.stringify(header), 'utf8')
  const parts: Buffer[] = [
    u32(GPU_IPC_MAGIC),
    u32(GPU_PROTOCOL_VERSION),
    u32(json.length),
    json,
    u32(blobs.size)
  ]
  for (const [name, data] of blobs) {
    const nameBuf = Buffer.from(name, 'utf8')
    parts.push(u32(nameBuf.length), nameBuf, u32(data.length), data)
  }
  return Buffer.concat(parts)
}

export function decodeGpuMessage(buf: Buffer): GpuIpcMessage {
  let off = 0
  const magic = buf.readUInt32LE(off)
  off += 4
  if (magic !== GPU_IPC_MAGIC) throw new Error('bad gpu ipc magic')
  const version = buf.readUInt32LE(off)
  off += 4
  if (version !== GPU_PROTOCOL_VERSION) throw new Error(`unsupported gpu ipc version ${version}`)
  const headerLen = buf.readUInt32LE(off)
  off += 4
  const header = JSON.parse(buf.subarray(off, off + headerLen).toString('utf8')) as unknown
  off += headerLen
  const nBlobs = buf.readUInt32LE(off)
  off += 4
  const blobs = new Map<string, Buffer>()
  for (let i = 0; i < nBlobs; i++) {
    const nameLen = buf.readUInt32LE(off)
    off += 4
    const name = buf.subarray(off, off + nameLen).toString('utf8')
    off += nameLen
    const dataLen = buf.readUInt32LE(off)
    off += 4
    blobs.set(name, Buffer.from(buf.subarray(off, off + dataLen)))
    off += dataLen
  }
  return { header, blobs }
}

export function readExact(fd: number, nbytes: number): Buffer {
  const buf = Buffer.alloc(nbytes)
  let off = 0
  while (off < nbytes) {
    const n = readSync(fd, buf, off, nbytes - off, null)
    if (n <= 0) throw new Error('gpu-engine closed the pipe')
    off += n
  }
  return buf
}

export function readGpuMessage(fd: number): GpuIpcMessage {
  const prefix = readExact(fd, 12)
  const magic = prefix.readUInt32LE(0)
  if (magic !== GPU_IPC_MAGIC) throw new Error('bad gpu ipc magic')
  const version = prefix.readUInt32LE(4)
  if (version !== GPU_PROTOCOL_VERSION) throw new Error(`unsupported gpu ipc version ${version}`)
  const headerLen = prefix.readUInt32LE(8)
  const headerBuf = readExact(fd, headerLen)
  const nBlobBuf = readExact(fd, 4)
  const nBlobs = nBlobBuf.readUInt32LE(0)
  const rest: Buffer[] = [prefix, headerBuf, nBlobBuf]
  for (let i = 0; i < nBlobs; i++) {
    const nameLen = readExact(fd, 4)
    const name = readExact(fd, nameLen.readUInt32LE(0))
    const dataLen = readExact(fd, 4)
    const data = readExact(fd, dataLen.readUInt32LE(0))
    rest.push(nameLen, name, dataLen, data)
  }
  return decodeGpuMessage(Buffer.concat(rest))
}

function u32(n: number): Buffer {
  const b = Buffer.alloc(4)
  b.writeUInt32LE(n >>> 0, 0)
  return b
}

File: gpu-engine/src/ipc.rs
use std::io::{self, Read, Write};

use serde_json::Value;

pub const MAGIC: u32 = 0x50464750;
pub const VERSION: u32 = 1;

pub struct Message {
    pub header: Value,
    pub blobs: Vec<(String, Vec<u8>)>,
}

fn read_u32<R: Read>(r: &mut R) -> io::Result<u32> {
    let mut buf = [0u8; 4];
    r.read_exact(&mut buf)?;
    Ok(u32::from_le_bytes(buf))
}

fn write_u32<W: Write>(w: &mut W, v: u32) -> io::Result<()> {
    w.write_all(&v.to_le_bytes())
}

pub fn write_message<W: Write>(w: &mut W, header: &Value, blobs: &[(String, Vec<u8>)]) -> io::Result<()> {
    let json = serde_json::to_vec(header).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
    write_u32(w, MAGIC)?;
    write_u32(w, VERSION)?;
    write_u32(w, json.len() as u32)?;
    w.write_all(&json)?;
    write_u32(w, blobs.len() as u32)?;
    for (name, data) in blobs {
        write_u32(w, name.len() as u32)?;
        w.write_all(name.as_bytes())?;
        write_u32(w, data.len() as u32)?;
        w.write_all(data)?;
    }
    w.flush()
}

pub fn read_message<R: Read>(r: &mut R) -> io::Result<Message> {
    let magic = read_u32(r)?;
    if magic != MAGIC {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "bad gpu ipc magic"));
    }
    let version = read_u32(r)?;
    if version != VERSION {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("unsupported gpu ipc version {version}"),
        ));
    }
    let header_len = read_u32(r)? as usize;
    let mut header_bytes = vec![0u8; header_len];
    r.read_exact(&mut header_bytes)?;
    let header: Value =
        serde_json::from_slice(&header_bytes).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
    let n_blobs = read_u32(r)? as usize;
    let mut blobs = Vec::with_capacity(n_blobs);
    for _ in 0..n_blobs {
        let name_len = read_u32(r)? as usize;
        let mut name = vec![0u8; name_len];
        r.read_exact(&mut name)?;
        let data_len = read_u32(r)? as usize;
        let mut data = vec![0u8; data_len];
        r.read_exact(&mut data)?;
        blobs.push((
            String::from_utf8(name).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?,
            data,
        ));
    }
    Ok(Message { header, blobs })
}
